src/utils/netutils.cpp

```diff
@@ -417,42 +472,52 @@ FILE* popen_conntrack(const char* source_ip, int source_port, int target_port)
     return popen(cmd, "r");
 }
 
-bool get_local_ip_address(IpAddress& client_address, int fd, const char **error_result)
+bool get_local_ip_address(IpAddress& client_address, int fd, const char **error_result) noexcept
 {
-    struct sockaddr_storage addr;
-    socklen_t namelen = sizeof(addr);
+    union
+    {
+        sockaddr s;
+        sockaddr_in s4;
+        sockaddr_in6 s6;
+        sockaddr_storage ss;
+    } u;
+    socklen_t namelen = sizeof(u);
     
-    if (::getsockname(fd, reinterpret_cast<sockaddr *>(&addr), &namelen) == -1)
+    std::memset(&u, 0, namelen);
+    if (::getsockname(fd, &u.s, &namelen) == -1)
     {
         if (error_result)
         {
             *error_result = "Cannot get local ip address";
         }
-        LOG(LOG_ERR, "getsockname failed with errno = %d (%s)", errno, strerror(errno));
+        LOG(LOG_ERR, "getsockname failed with errno = %d (%s)",
+            errno,
+            strerror(errno));
         return false;
     }
     else
     {
-        const char *res = nullptr; 
-        
-        res = (addr.ss_family == AF_INET) ?
-            inet_ntop(AF_INET,
-                      &reinterpret_cast<sockaddr_in *>(&addr)->sin_addr,
-                      client_address.ip_addr,
-                      INET_ADDRSTRLEN) :
-            inet_ntop(AF_INET6,
-                      &reinterpret_cast<sockaddr_in6 *>(&addr)->sin6_addr,
-                      client_address.ip_addr,
-                      INET6_ADDRSTRLEN);
-        if (!res)
+        char ip_address[NI_MAXHOST] { };
+
+        if (int res = ::getnameinfo(&u.s,
+                                    sizeof(u.ss),
+                                    ip_address,
```

Comment:
should be `client_address.ip_addr`.

Reply:
Yes, it will avoid temporary buffer + O(n) with strncpy